demos/nonlinear_QG_winddrivengyre/qg_winddrivengyre.py.rst

```diff
@@ -0,0 +1,279 @@
+
+  phi, psi = TestFunction(Vcg), TrialFunction(Vcg)
+
+We must define functions that will store our linear and nonlinear solutions.
+In order to solve the nonlinear problem, we use the linear
+solution as a guess for the nonlinear problem. ::
+
+  psi_lin = Function(Vcg, name='Linear Streamfunction')
+  psi_non = Function(Vcg, name='Nonlinear Streamfunction')
+
+We can finally write down the linear Stommel equation in its weak
+form. We will use the solution to this as the input for the nonlinear
+Stommel equation. ::
+
+  a = - r * inner(grad(psi), grad(phi)) * dx - F * psi * phi * dx + beta * psi.dx(0) * phi * dx
+  L = Qwinds * phi * dx
+
+We set-up an elliptic solver for this problem, and solve for the
+linear streamfunction, ::
+
+  linear_problem = LinearVariationalProblem(a, L, psi_lin, bcs=bc)
+  linear_solver = LinearVariationalSolver(linear_problem,
+                                          solver_parameters= {'ksp_type': 'preonly',
+                                                              'pc_type': 'lu'})
+  linear_solver.solve()
+
+We will redefine the nonlinear stream-function as its guess, the
+linear stream-function ::
+
+  psi_non.assign(psi_lin)
+
+And now we can define the weak form of the nonlinear problem. Note
+that the problem is stated in residual form so there is no trial function. ::
+
+  G = - inner(grad(phi), perp(grad(psi_non))) * div(grad(psi_non)) * dx \
+      -r * inner(grad(psi_non), grad(phi)) * dx - F * psi_non * phi * dx \
+      + beta * psi_non.dx(0) * phi * dx \
+      - Qwinds * phi * dx
+
+We solve for the nonlinear streamfunction now by setting up another
+elliptic solver, ::
+
+  nonlinear_problem = NonlinearVariationalProblem(G, psi_non, bcs=bc)
+  nonlinear_solver = NonlinearVariationalSolver(nonlinear_problem,
+                                                solver_parameters= {'snes_type': 'newtonls',
+                                                                    'ksp_type': 'preonly',
+                                                                    'pc_type': 'lu'})
+  nonlinear_solver.solve()
+
+Now that we have the full solution to the nonlinear Stommel problem,
+we can plot it, ::
+
+  try:
+      import matplotlib.pyplot as plt
+  except:
+      warning("Matplotlib not imported")
+
+  try:
+      plot(psi_non)
+  except Exception as e:
+      warning("Cannot plot figure. Error msg '%s'" % e.message)
+
+  try:
+      plt.show()
+  except Exception as e:
+      warning("Cannot show figure. Error msg '%s'" % e.message)
+
+  file = File('Nonlinear Streamfunction.pvd')
+  file.write(psi_non)
+
+We can also see the difference between the linear solution and the
+nonlinear solution. We do this by defining a weak form.  (Note: other approaches may be possible.) ::
+
+  tf, difference = TestFunction(Vcg), TrialFunction(Vcg)
+
+  a = difference * tf * dx
+  L = (psi_lin - psi_non) * tf * dx
+  difference = Function(Vcg, name='Difference')
+  solve(a == L, difference, None)
```

Comment:
This is a projection of the difference, since `psi_lin` and `psi_non` are in the same space, we could instead just write:
`difference = assemble(psi_lin - psi_non)`